cmd/cellrangerBam/cellrangerBam.go

```diff
@@ -0,0 +1,823 @@
+func umiSaturation(umiBxSlice []string, out *fileio.EasyWriter) {
+	var perc, randNum float64
+	var j string
+	var count int
+
+	fileio.WriteToFileHandle(out, "totalReads\tumis")
+
+	for i := 1; i <= 10; i++ {
+		perc = float64(i) / 10
+		mp := make(map[string]bool)
+		count = 0
+		for _, j = range umiBxSlice {
+			randNum = rand.Float64()
+			if randNum > perc {
+				continue
+			}
+			count++
+			mp[j] = true
+		}
+		fileio.WriteToFileHandle(out, fmt.Sprintf("%d\t%d", count, len(mp)))
+	}
+}
+
+// combineBams is triggered when there are more than 1 input bams in a comma delimited list. This function appends a "_int" to the back of the cell barcode correstponding to the input file order
+// and then concatenates all bam files together
+func combineBams(a string, out *fileio.EasyWriter, iteration *int, length int, bw *sam.BamWriter) *sam.BamWriter {
+	var err error
+	var columns, columns2 []string
+	var j, join string
+	var n int
+
+	inChan, head := sam.GoReadToChan(a)
+	if *iteration == 1 {
+		bw = sam.NewBamWriter(out, head)
+	}
+	for i := range inChan {
+		err = sam.ParseExtra(&i)
+		columns = strings.Split(i.Extra, "\t")
+		for n, j = range columns {
+			columns2 = strings.Split(j, ":")
+			if columns2[0] == "CB" {
+				columns[n] = fmt.Sprintf("%s_%d", j, *iteration)
+				join = strings.Join(columns, "\t")
+				i.Extra = join
+			}
+		}
+		sam.WriteToBamFileHandle(bw, i, 0)
+	}
+	exception.PanicOnErr(err)
+	*iteration++
+	return bw
+}
+
+// binnedPseudobulk is the psuedobulk function if the -binCells option is used. It adds an addition column to the dataframe corresponding to bin identity
+func binnedPseudobulk(inSlices [][]string, out *fileio.EasyWriter, norm string) {
+	var j, i string
+	var count float64
+	var found bool
+	var columns []string
+
+	whichBin := 'A'
+	fileio.WriteToFileHandle(out, "construct\tcounts\tbin")
+	for _, bin := range inSlices {
+		mp := make(map[string]float64)
+		var toWrite []string
+		for _, j = range bin {
+			columns = strings.Split(j, "\t")
+			count, found = mp[columns[1]]
+			if !found {
+				mp[columns[1]] = 1
+			} else {
+				mp[columns[1]] = count + 1
+			}
+		}
+		if norm != "" {
+			inputNormalize(mp, norm)
+		}
+		for i = range mp {
+			toWrite = append(toWrite, fmt.Sprintf("%s\t%f\t%c", i, mp[i], whichBin))
+		}
+		sort.Strings(toWrite)
+		for _, i = range toWrite {
+			fileio.WriteToFileHandle(out, i)
+		}
+		whichBin++
+	}
+}
+
+// determine ideal bins will determine how many bins should be used for the pseudobulk -binCells option
+func determineIdealBins(s Settings, cellTypeSlice []string) int {
+	var count, j int
+	var found, stop bool
+	var z, l, m string
+	var binSlice []int
+	var matrix [][]string
+
+	nc := fileio.Read(s.determineBins)
+
+	for i := 0; i <= 10; i++ { //do this whole loop 10 times (essentially 10 replicates)
```

Comment:
I think this loop is done 11 times because it gets executed for i==0,1,2,...10

Reply:
haha good catch